File: services/api/app/schemas/spending.py
from datetime import date as date_type
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class CategorySpend(BaseModel):
    category_id: UUID | None = None
    name: str
    icon: str | None = None
    color: str | None = None
    amount: Decimal
    count: int
    budget: Decimal | None = None


class DailySpend(BaseModel):
    date: date_type
    amount: Decimal


class MonthSpend(BaseModel):
    month: str  # YYYY-MM
    amount: Decimal


class MerchantSpend(BaseModel):
    name: str
    amount: Decimal
    count: int


class CurrencyTotal(BaseModel):
    currency: str
    amount: Decimal


class SpendingSummary(BaseModel):
    """
    The user's spending for one month. "Spending" is the user's own share:
    the full amount of personal expenses plus their split of group expenses.
    Totals only count expenses in `currency`; others land in `other_currencies`.
    """

    month: str
    currency: str
    total: Decimal
    previous_total: Decimal
    expense_count: int
    personal_total: Decimal
    group_share_total: Decimal
    overall_budget: Decimal | None = None
    by_category: list[CategorySpend] = Field(default_factory=list)
    daily: list[DailySpend] = Field(default_factory=list)
    trend: list[MonthSpend] = Field(default_factory=list)
    top_merchants: list[MerchantSpend] = Field(default_factory=list)
    other_currencies: list[CurrencyTotal] = Field(default_factory=list)
    income_total: Decimal
    net: Decimal

File: apps/web/src/types/spending.ts
/** Money fields are decimal strings, like everywhere else in the API. */
export interface CategorySpend {
  category_id: string | null;
  name: string;
  icon: string | null;
  color: string | null;
  amount: string;
  count: number;
  budget: string | null;
}

/**
 * One month of the user's own spending: the full amount of personal expenses
 * plus their share of group expenses, in their default currency only.
 */
export interface SpendingSummary {
  /** YYYY-MM */
  month: string;
  currency: string;
  total: string;
  previous_total: string;
  expense_count: number;
  personal_total: string;
  group_share_total: string;
  overall_budget: string | null;
  by_category: CategorySpend[];
  daily: { date: string; amount: string }[];
  trend: { month: string; amount: string }[];
  top_merchants: { name: string; amount: string; count: number }[];
  other_currencies: { currency: string; amount: string }[];
  income_total: string;
  net: string;
}
